src/java/org/apache/cassandra/db/rows/Cell.java

```diff
@@ -59,16 +70,37 @@
 
     public interface Factory<V>
     {
-        Cell<V> create(ColumnMetadata column, long timestamp, int ttl, int localDeletionTime, V value, CellPath path);
+        Cell<V> create(ColumnMetadata column, long timestamp, int ttl, long localDeletionTime, V value, CellPath path);
     }
 
     protected Cell(ColumnMetadata column)
     {
         super(column);
     }
 
+    public static int deletionTimeLongToUnsignedInteger(long deletionTime)
+    {
+        return deletionTime == NO_DELETION_TIME ? NO_DELETION_TIME_UNSIGNED_INTEGER : CassandraUInt.fromLong(deletionTime);
+    }
+
+    public static long deletionTimeUnsignedIntegerToLong(int deletionTimeUnsignedInteger)
+    {
+        return deletionTimeUnsignedInteger == NO_DELETION_TIME_UNSIGNED_INTEGER ? NO_DELETION_TIME : CassandraUInt.toLong(deletionTimeUnsignedInteger);
+    }
+
+    public static long getVersionedMaxDeletiontionTime()
+    {
+        if (BigFormat.ttlMode.equals(TTL_MODE.EXTENDED_TTL_2106))
+            // The whole cluster is 2016, we're out of the 2038/2106 mixed clusters scenario. Short cut to avoid the 'minClusterVersion' volatile read
```

Comment:
Does the distinction between `COMPATIBILITY` and `EXTENDED_TTL_2106` exist only to avoid the cost of reading a `volatile` while checking the cluster version? I doubt that reading yet another `volatile` is going to have a lot of impact. 

If the problem is also not trusting the cluster version, I think that the situation is no different to what happens during any other operation done during a rolling upgrade, independently of the extended expiration dates. IMO that problem should be addressed by CASSANDRA-18301, and here we can assume that announcing the specified messaging version is enough. I mean, setting the messaging version with the property makes this identical to any other operation during a rolling upgrade.

So I think the property could be a boolean. Going a step beyond, the property could set the major Cassandra version we want to stay compatible with. That would determine the sstable, commitlog, hints and messaging versions, as well as the availability of the extended TTLs. That is quite easy to do given the excellent state of this patch. It has the advantage of making things a bit easier for CASSANDRA-18301. It also breaks the odd dependency on `BigFormat` that the cells, messaging, commitlog and hints currently have. 

I gave it a go in this commit: https://github.com/adelapena/cassandra/commit/0a87f1ee13e5eed92dee637638667a30695c3426

It replaces the `cassandra.sstable.format.ttl_mode` enum property by a more generic `cassandra.storage_compatibility_version` int property. That property is meant to determine the entire adoption of not-backwards compatible features. 

As for the cost of checking the `volatile` in `Cell.getVersionedMaxDeletiontionTime`, I think we can do an optimization to reduce its cost in the common case. Most of the callers of that method are interested only in checking is a certain timestamp exceeds it. And most of the checked timestamps will be, I assume, under both the 2038 and 2106 limit. So we can add a `Cell.exceedsVersionMaxDeletionTime` method that returns `false` if the value is lesser than both caps without checking the cluster version. This way: https://github.com/adelapena/cassandra/commit/ae0941014bf32cfd471bc5b8d57fb27b4365b959

Reply:
The distinction between `COMPATIBILITY` and `EXTENDED_TTL_2106` does spare the extra volatile read yes. But also 'freezes' the status of the cluster. If sbdy starts a 4.0 node by accident or similar things happen the nodes won't be flapping from on behavior to the other which sounds very dangerous imo. With `EXTENDED_TTL_2106` only the accidental node will suffer and the rest of the cluster will be just fine.

I think your proposal for the avoidance of the volatile read, iiuc, will always hit the volatile when you're >=5.0 and using MAX_TTL (or anything going beyond 2038) which is quite a common scenario and it will be more common as we approach 2038.

Reply:
Agree on both things. Trying to infer the behaviour should only be done during upgrades, and the avoidance on the version read will miss the `MAX_TTL` case.

However, I still think that the property can easily be more generic, so we don't merge code that will need to be undone by [CASSANDRA-18301](https://issues.apache.org/jira/browse/CASSANDRA-18301). The compatibility mode property can be generalized to cover any similar features with minimal changes. It can also be put on its own class and made accessible through `DatabaseDescriptor`, so we don't overload `BigFormat` nor we have messaging, commitlog, hints, etc. depending on that particular sstable format.

I gave it a go here: https://github.com/adelapena/cassandra/commit/bfbbb449b2645cef80d7ed040593ff07492610ec

It's basically the same approach but with some minor refactoring and a new rolling upgrade test.

Reply:
Thx for that commit. I have shamelessly stolen it with some fixed test failures. Some comments:
- I like the 'generic' spin you gave it which will be nice for 18301
- I like it respects the tri-state. Here we're avoiding a volatile read. But if 18301 goes TCM or anything else, 99% probability it will be more expensive than a volatile read, we need a mechanism to avoid that cost in the interim upgrading state.
- I like it respects the idea of signaling the upgrade having finished and not having a window of opportunity to go back/forth accidentally.